### ISSUE TITLE
Conversion vs. validation

### ISSUE THREAD

**Issue by lshabbot** [schematics/schematics#137]
Just curious if this is as design.  When I use a DateTimeType, I was expecting to get a  ValidationError if the time format is incorrect, but instead I am getting a ModelConversionError.  I was hoping to get all possible validation errors when validating a resource, but based on this implementation, if the time format is incorrect, it will fail immediately before validating any other fields.  I guess a work-around would be to define the type as StringType and define my on validator??

**Comment by chekunkov** [comment]
+1 to this question

**Comment by bintoro** [comment]
Can someone explain to me in clear terms what the issue is? I'm getting error messages for all applicable fields, both when converting and when validating.

```python
class M(Model):

    datefield = DateTimeType(required=True)
    intfield = IntType(min_value=9)
```

Trying to import values that cannot be interpreted in terms of the specified types:

```python
>>> m = M({'datefield':'impossible_value', 'intfield':'impossible_value'})
schematics.exceptions.ModelConversionError: {
  'datefield': [u'Could not parse impossible_value. Should be ISO8601.'], 
  'intfield': [u"Value 'impossible_value' is not int."]
}
```

Trying to validate values that do not satisfy validity conditions:

```python
>>> m = M()
>>> m.datefield = None
>>> m.intfield = 5
>>> m.validate()
schematics.exceptions.ModelValidationError: {
  'datefield': [u'This field is required.'], 
  'intfield': [u'Int value should be greater than 9.']
}
```

**Comment by chekunkov** [comment]
This library is new for me and TBH I don't understand why it separates parsing errors and validation errors. I would love to hear why this design decision was made. Here's an example:

```python
from schematics.models import Model
from schematics.types import IntType, DateType
from schematics.types.compound import ListType, ModelType


class ObjectModel(Model):
    id = IntType(min_value=0, required=True)
    date = DateType(required=True)


class CollectionModel(Model):
    data = ListType(ModelType(ObjectModel), min_size=1, required=True)


def test1():
    """Passes."""
    return CollectionModel({'data': [
        {'id': -1},
        {'id': 10}
    ]})


def test1_1():
    """Fails with::

        ModelValidationError: {'data': [
            {'date': ['This field is required.'],
             'id': ['Int value should be greater than 0.']},
            {'date': ['This field is required.']}
        ]}

    """
    model = test1()
    model.validate()


def test2():
    """Fails with::

        ModelConversionError: {'data':
            {'date': ['Could not parse aaa. Should be ISO8601 (YYYY-MM-DD).']}
        }

    """
    return CollectionModel({'data': [
        {'id': -1, 'date': 'aaa'},
        {'id': 10}
    ]})


def test2_1():
    """Fails because of test2() failure.
    """
    model = test2()
    model.validate()
```

The only thing that changes is date field for one of the nested objects - but see how behaviour changes in this case. Instead I would expect test2() to pass and test2_1() to fail with validation error like:

```
{'data': [
    {'date': ['Could not parse aaa. Should be ISO8601 (YYYY-MM-DD).'],
     'id': ['Int value should be greater than 0.']},
    {'date': ['This field is required.']}
]}
```

Another thing that confuses me. Models allow 2 ways of passing values to them - pass raw data in `__init__` and by setting attributes. Raw data example is above. And here's how model behaves when one is setting attribute values:

```python
collection_model = CollectionModel()
# hm, no ModelConversionError (parsing error)
collection_model.data = [{'id': -1, 'date': 'aaa'}, {'id': 10}]
collection_model.validate()
```

This snippet raises

```
ModelValidationError: {'data': {'date': ['Could not parse aaa. Should be ISO8601 (YYYY-MM-DD).']}}
```

So previously that was `ModelConversionError`, and it wasn't possible to create model with incorrect values types (I believe that's the main purpose of parse/validation errors separation) - now it raises different error type and model does contain wrong value types before validation step.

And even more interesting example:

```python
object_model = ObjectModel()
object_model.date = 'aaa'
object_model.id = -1
object_model.validate()
```

Raises

```
ModelValidationError: {'date': ['Could not parse aaa. Should be ISO8601 (YYYY-MM-DD).'], 'id': ['Int value should be greater than 0.']}
```

In contrast to `test2()` here `ModelValidationError` mixes both validation errors ('Int value should be greater than 0.') and type errors ('Could not parse aaa. Should be ISO8601 (YYYY-MM-DD).').

So this confuses me a lot - I would prefer to have consistent behaviour and here it looks like model parsing/validation behaves differently based on model structure and how one fill the model with data.

**Comment by bintoro** [comment]
Thanks for the great writeup, @chekunkov! There are definitely a few things we could do to improve these aspects.

#### Separation between parsing and validation

I can't speak for the original authors, but I suppose the motivation behind separating parsing from validation is flexibility. Basically, you may want to work with objects that don't satisfy validity conditions right off the bat, which would get cumbersome and inefficient if the import step were to run validators automatically.

Also, parsing errors (inherent to the input) and validity errors (logical constraints that may change) are fundamentally different, so IMO it's a distinction worth preserving.

That said, I don't see why we couldn't make things simpler for when we don't care about the aforementioned distinction. For example, `instance = SomeModel(<some data>, validate=True)` might run validators automatically and give you all the errors right away.

Interestingly, `required=True` is a validator that is already recognized at the import step, but the relevant switch `partial` is missing from `Model.__init__`! The idea is that doing `instance = SomeModel(<some data>, partial=False)` should complain about parsing errors *and* missing fields but not other validity problems. You can already do this by creating an empty instance and importing separately: `instance.import_data(<data>, partial=False)`.

#### Exceptions

As you've discovered, the exception you get is solely dependent on the process: upon import a `ModelConversionError` is raised, while the validation step raises `ModelValidationError`, even if the underlying error is the same.

I suspect this is the source of most of the confusion here.

Having these two different exceptions seems completely useless and misleading to me too. It would make more sense to separate errors into *parsing errors* and *validity errors*. But obviously these cannot correspond to Python exceptions, since the same validation process can yield both kinds of errors.

I suppose that, ideally, there would be just one kind of exception, and the information about the type of the problem would be contained in the message object. That is, for each offending field, you'd get (1) the name of the field, (2) the type of the problem (parser/validator), and (3) the message.

#### Unparseable data in instances

You can indeed assign completely spurious data directly to instance attributes. While it certainly would be possible to call an import/validation routine right there and then, it would mean that every simple assignment operation would have to be enclosed in a try-except statement:

```python
try:
    collection_model.data = [{'id': -1, 'date': 'aaa'}, {'id': 10}]
except ModelConversionError:
    # handle the error; probably store the info somewhere while you set other fields
```

It seems way more practical to be able to manipulate instance attributes at will, and only once you're done, *then* validate everything in one fell swoop instead of manually try-excepting and collecting errors. Besides, you can always do this, which will raise a `ModelConversionError` immediately:

```python
collection_model.import_data({'data': [{'id': -1, 'date': 'aaa'}, {'id': 10}]})
```

But again, I think the `ModelConversionError/ModelValidationError` weirdness may be the primary culprit here, making things seem illogical. Would you agree or am I just not grokking something?

I mean, given that

  * validating implies parsing but not the other way around
  * the two exceptions don't really have any difference in meaning

none of the test cases seem to behave in a surprising way.

Thoughts?

**Comment by chekunkov** [comment]
> Also, parsing errors (inherent to the input) and validity errors (logical constraints that may change) are fundamentally different, so IMO it's a distinction worth preserving.

Agree, but this needs some clarification on what should be treated as parsing error and what should be validity errors. E.g. currently incorrect date format is treated as parsing error. I'm not sure that's correct approach. Let me explain. Obviously there are 2 steps: import and validation.

1\. Import. IMO the only thing that should be validated during this step is 'shape' of data. I.e. detect rogue fields and check required fields (if `partial=False`) then follow nested fields and check the same for nested objects. As far as data does not satisfy shape constraints - it doesn't make sense to validate it - because model level validation often requires several fields to be present in expected positions. This is the case where separation between parsing and validation errors shines.

2\. Validation. Here any other constraints should be checked - type errors, format errors, 'greater than zero' constraints like. And ideally there should be 2 modes - 'fail early' - give first occurred validation error and 'show all' - give list with all errors.

I'm not saying it's the only right way, but that's how I see it.

Given model from your comment:

```
m = M({'datefield':'impossible_value', 'intfield': 5})
```

should pass

```
m.validate()
```

should give 

```
schematics.exceptions.ModelValidationError: {
  'datefield': [u'Could not parse impossible_value. Should be ISO8601.'], 
  'intfield': [u'Int value should be greater than 9.']
}
```

```
m = M({'intfield': 5})
```
should fail with

```
schematics.exceptions.ModelValidationError: {
  'datefield': [u'This field is required.'],
}
```

```
m = M({'blah': 5})
```
should fail with 

```
schematics.exceptions.ModelValidationError: {
  'datefield': [u'This field is required.'],
  'blah': [u'Rogue field']
}
```

Unfortunately it doesn't work like that now. In case of DateType - parse error occurs even if you're sending date in incorrect format or breaking some format constraints although shape of the object is correct, datefield is of the string type and only breaks month limit.

```
>>> M({'intfield': '1', 'datefield': '2015-13-12'})
ModelConversionError: {'datefield': ['Could not parse 2015-13-12. Should be ISO8601 (YYYY-MM-DD).']}
```

**Comment by bintoro** [comment]
> Agree, but this needs some clarification on what should be treated as parsing error and what should be validity errors.

This is actually well-defined. The conversion step is about turning the *primitive representation* of every field to a *native representation* by calling `to_native()` on each respective type. After a successful data import, you're guaranteed to end up with a set of values with predefined Python types. It's not just about the shape of the containing model object.

For instance, the primitive representation of a `DateType` is a string, while the native form is a Python `datetime.date` object. After a successful conversion from raw data, you're guaranteed to have a `date` object to work with.

There is absolutely no way '2015-13-12' and 'impossible_value' can be reliably interpreted as dates in the Gregorian calendar and, consequently, as `date` objects. Therefore they must be rejected at the conversion stage.

**Comment by bintoro** [comment]
I happened to accidentally find some interesting historical discussion on this very design choice, have a look: https://github.com/plain-vanilla-games/schematics/pull/11.

**Comment by lkraider** [comment]
Conversion exists so you get a reliable type when accessing the attributes.

Think of conversion as a type cast - you can set an `Integer` field with `sys.max_int`, but not with `"a number"`. Validation is a further step that restricts the type, so for example you can restrict your `Integer` field to be no more that `1024`, so that `sys.max_int` fails validation.

Formally, conversion is a function that returns a type, while validation is a function that returns a boolean. Causing exception on conversion is a requirement, while raising exception on validation is a matter of convention.

We could perform late/lazy conversion, but seeing no further step can be done without it, makes sense to perform on the model init - and also because otherwise you would get an instance that causes exceptions on attribute access.

@bintoro: I have been running my own branch in production from those days exactly because I need those integrity guarantees, summed up in #124 :

> This guarantees mean:
> 
> * always serialize ONLY the valid data of a model
> * don't loose current valid data when setting invalid data

In my view the split of _input data_ vs _valid data_ is a requirement of correctness, which I tried to describe mathematically in https://github.com/nKey/schematics/tree/doc-math/docs/math (my first stab at trying to describe an algorithm using a formal model).

**Comment by bintoro** [comment]
@lkraider I've looked at #124 before and very much like the idea that you can muck around with an instance without losing what's already been validated. I think Schematics should have that functionality.

The Undefined feature will necessary involve a rethink about `_data` and its parity with `_fields`. So I think the time is ripe to finally bring this idea aboard.

**Comment by bintoro** [comment]
With #374 now merged into development, it's possible to perform conversion and validation in one fell swoop.

The confusing separation of aggregate errors into `ModelValidationError` and `ModelConversionError` is gone. The distinction now only exists at the field level where it makes sense, while aggregate errors are always delivered as a `DataError` instance. The individual `ConversionError` and `ValidationError` instances are preserved in the `DataError` message structure.
